mypy/checker.py

```diff
@@ -1715,15 +1797,32 @@ def check_incompatible_property_override(self, e: Decorator) -> None:
 
     def visit_with_stmt(self, s: WithStmt) -> Type:
         echk = self.expr_checker
+        if s.is_async:
+            m_enter = '__aenter__'
+            m_exit = '__aexit__'
+        else:
+            m_enter = '__enter__'
+            m_exit = '__exit__'
         for expr, target in zip(s.expr, s.target):
             ctx = self.accept(expr)
-            enter = echk.analyze_external_member_access('__enter__', ctx, expr)
+            enter = echk.analyze_external_member_access(m_enter, ctx, expr)
             obj = echk.check_call(enter, [], [], expr)[0]
+            if s.is_async:
+                self.check_subtype(obj, self.named_type('typing.Awaitable'), expr)
             if target:
+                if s.is_async:
+                    # XXX TODO What if it's a subclass of Awaitable?
+                    if (isinstance(obj, Instance) and len(obj.args) == 1
+                            and obj.type.fullname() == 'typing.Awaitable'):
+                        obj = obj.args[0]
+                    else:
+                        obj = AnyType()
                 self.check_assignment(target, self.temp_node(obj, expr))
-            exit = echk.analyze_external_member_access('__exit__', ctx, expr)
+            exit = echk.analyze_external_member_access(m_exit, ctx, expr)
             arg = self.temp_node(AnyType(), expr)
-            echk.check_call(exit, [arg] * 3, [nodes.ARG_POS] * 3, expr)
+            res = echk.check_call(exit, [arg] * 3, [nodes.ARG_POS] * 3, expr)[0]
+            if s.is_async:
+                self.check_subtype(res, self.named_type('typing.Awaitable'), expr)
         self.accept(s.body)
```

Comment:
I'm not certain, but I think it might be a bit clearer overall to separate these changes out into a different method like you did for async for.